### PR TITLE
gh-91049: Introduce set vectorcall field API for PyFunctionObject

### PR DESCRIPTION
**Pull request by adphrost**
Avoid specializing functions with overridden vectorcall field

* Issue: gh-91049

### CODE REVIEW

**Comment by adphrost** [comment]
@markshannon Can you let me know if there should be any other checks in `ceval.c` besides the one added in this PR?

For context, this check fixes an edge case where calling function with a set vectorfield only worked with a unpacked tuple, e.g. f`(1)` failed but `f(*(1,))` succeeded (now both succeed)

File: Lib/test/test_call.py
Comment (itamaro) [inline Lib/test/test_call.py]:
```suggestion
        self.assertEqual(11, f(num))
```

File: Lib/test/test_call.py
Comment (itamaro) [inline Lib/test/test_call.py]:
```suggestion
            self.assertIsNone(f(num))
```

File: Python/ceval.c
Comment (itamaro) [inline Python/ceval.c]: this probably exceeds line length

**Review by itamaro — COMMENTED** [review]
thanks for working on this @adphrost !

File: Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst]: Please document the function in Doc/c-api/function.rst and in the NEWS entry, refer to use with:

```
:c:func:`PyFunction_SetVectorcall`
```

New public functions should also be documented in What's New in Python 3.11 > C API > New features: Doc/whatsnew/3.11.rst.

File: Objects/funcobject.c
Comment (markshannon) [inline Objects/funcobject.c]: This can't fail. Return `void`

File: Modules/_testcapimodule.c
Comment (markshannon) [inline Modules/_testcapimodule.c]: The names should match, and this only applies to functions. I'd suggest `function_setvectorcall`

File: Modules/_testcapimodule.c
Comment (markshannon) [inline Modules/_testcapimodule.c]: You can use `METH_O`. You'll also need to check that `func` is a function.

**Review by markshannon — COMMENTED** [review]
A few comments.
Have you attempted to measure the performance impact of this? (I'd expect it to be small or negligible).

File: Python/ceval.c
Comment (adphrost) [reply Python/ceval.c]: What is the max line length allowed? And is there a linter I can run to handle this for me?

**Comment by itamaro** [comment]
> Have you attempted to measure the performance impact of this? (I'd expect it to be small or negligible).

@markshannon is there a simple way to ask a buildbot to run pyperformance comparing this against the base revision? we can do that manually, but a buildbot doing it for us would be nice :)

**Comment by itamaro** [comment]
circling back to @markshannon 's performance impact question. -- tldr `Geometric mean: 1.00x faster`

some more details:

rebased this PR onto a4460f2eb8 and ran pyperformance on a4460f2eb8 and on the rebased branch

both baseline and branch were built using `./configure --enable-optimizations --with-lto && make -j`

used pyperformance 1.0.5 (installed directly from GitHub) on bare metal AWS machine (Linux-5.13.0-1023-aws-x86_64-with-glibc2.31 with 72 logical CPUs), and excluded the sqlalchemy benchmarks because they tried building greenlet and failed

```
pyperformance run --benchmarks=-sqlalchemy_declarative,-sqlalchemy_imperative --python builds/3.12-opt/python -o main.json --rigorous
pyperformance run --benchmarks=-sqlalchemy_declarative,-sqlalchemy_imperative --python builds/3.12-gh-91049-vectorcall-opt/python -o gh-91049.json --rigorous
```

compare results:

```
pyperf compare_to main.json gh-91049.json -G
Slower (20):
- regex_effbot: 3.02 ms +- 0.01 ms -> 3.17 ms +- 0.01 ms: 1.05x slower
- meteor_contest: 110 ms +- 2 ms -> 114 ms +- 4 ms: 1.04x slower
- pidigits: 200 ms +- 0 ms -> 208 ms +- 0 ms: 1.04x slower
- telco: 7.17 ms +- 0.21 ms -> 7.39 ms +- 0.34 ms: 1.03x slower
- html5lib: 64.0 ms +- 2.7 ms -> 65.6 ms +- 2.6 ms: 1.02x slower
- unpickle_list: 5.30 us +- 0.08 us -> 5.40 us +- 0.05 us: 1.02x slower
- pyflate: 430 ms +- 3 ms -> 438 ms +- 4 ms: 1.02x slower
- sympy_sum: 168 ms +- 2 ms -> 171 ms +- 2 ms: 1.02x slower
- xml_etree_iterparse: 110 ms +- 1 ms -> 112 ms +- 3 ms: 1.02x slower
- async_tree_memoization: 640 ms +- 25 ms -> 650 ms +- 25 ms: 1.01x slower
- chaos: 71.8 ms +- 0.7 ms -> 72.8 ms +- 0.6 ms: 1.01x slower
- hexiom: 6.71 ms +- 0.06 ms -> 6.80 ms +- 0.06 ms: 1.01x slower
- sympy_str: 299 ms +- 3 ms -> 302 ms +- 4 ms: 1.01x slower
- spectral_norm: 101 ms +- 1 ms -> 103 ms +- 1 ms: 1.01x slower
- sympy_integrate: 21.5 ms +- 0.2 ms -> 21.8 ms +- 0.2 ms: 1.01x slower
- logging_silent: 105 ns +- 1 ns -> 107 ns +- 3 ns: 1.01x slower
- raytrace: 303 ms +- 2 ms -> 306 ms +- 4 ms: 1.01x slower
- regex_v8: 23.1 ms +- 0.3 ms -> 23.3 ms +- 0.1 ms: 1.01x slower
- unpickle_pure_python: 239 us +- 2 us -> 240 us +- 2 us: 1.01x slower
- 2to3: 282 ms +- 2 ms -> 283 ms +- 1 ms: 1.00x slower

Faster (25):
- genshi_xml: 57.1 ms +- 2.7 ms -> 54.3 ms +- 1.8 ms: 1.05x faster
- scimark_sparse_mat_mult: 4.93 ms +- 0.06 ms -> 4.70 ms +- 0.10 ms: 1.05x faster
- scimark_lu: 112 ms +- 2 ms -> 109 ms +- 2 ms: 1.03x faster
- json_dumps: 13.3 ms +- 0.2 ms -> 12.9 ms +- 0.1 ms: 1.03x faster
- scimark_fft: 337 ms +- 3 ms -> 331 ms +- 3 ms: 1.02x faster
- nbody: 97.6 ms +- 1.8 ms -> 95.9 ms +- 1.8 ms: 1.02x faster
- pathlib: 23.0 ms +- 1.2 ms -> 22.6 ms +- 0.8 ms: 1.02x faster
- pickle_pure_python: 324 us +- 3 us -> 318 us +- 3 us: 1.02x faster
- genshi_text: 23.2 ms +- 0.4 ms -> 22.8 ms +- 0.3 ms: 1.02x faster
- unpack_sequence: 46.5 ns +- 0.6 ns -> 45.7 ns +- 0.7 ns: 1.02x faster
- fannkuch: 401 ms +- 6 ms -> 396 ms +- 4 ms: 1.01x faster
- chameleon: 7.06 ms +- 0.06 ms -> 6.98 ms +- 0.08 ms: 1.01x faster
- pickle_dict: 27.5 us +- 0.1 us -> 27.2 us +- 1.0 us: 1.01x faster
- pickle: 10.2 us +- 0.1 us -> 10.1 us +- 0.1 us: 1.01x faster
- xml_etree_parse: 170 ms +- 2 ms -> 168 ms +- 3 ms: 1.01x faster
- xml_etree_process: 56.8 ms +- 0.5 ms -> 56.3 ms +- 0.5 ms: 1.01x faster
- nqueens: 88.0 ms +- 1.1 ms -> 87.2 ms +- 1.1 ms: 1.01x faster
- json_loads: 26.3 us +- 0.3 us -> 26.1 us +- 0.2 us: 1.01x faster
- async_tree_none: 549 ms +- 15 ms -> 544 ms +- 12 ms: 1.01x faster
- async_tree_cpu_io_mixed: 771 ms +- 13 ms -> 766 ms +- 10 ms: 1.01x faster
- mako: 10.5 ms +- 0.0 ms -> 10.4 ms +- 0.1 ms: 1.01x faster
- crypto_pyaes: 77.2 ms +- 0.8 ms -> 76.8 ms +- 0.9 ms: 1.01x faster
- xml_etree_generate: 81.2 ms +- 0.6 ms -> 80.8 ms +- 0.5 ms: 1.00x faster
- regex_compile: 142 ms +- 1 ms -> 141 ms +- 1 ms: 1.00x faster
- scimark_sor: 121 ms +- 1 ms -> 120 ms +- 1 ms: 1.00x faster

Benchmark hidden because not significant (18): async_tree_io, deltablue, django_template, dulwich_log, float, go, logging_format, logging_simple, pickle_list, python_startup, python_startup_no_site, regex_dna, richards, scimark_monte_carlo, sqlite_synth, sympy_expand, tornado_http, unpickle

Geometric mean: 1.00x faster
```

thanks @ansley for doing the perf analysis on this :)

File: Doc/c-api/function.rst
Comment (vstinner) [inline Doc/c-api/function.rst]:
```suggestion
   Set the vectorcall field of a given function object *func*.
```

Please add ``.. versionadded:: 3.12``.

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c]: Could you modify the test to return the string "overridden" instead of just returning None?

File: Doc/whatsnew/3.12.rst
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]: You may explain the purpose of the function here, like mention Cinder JIT and Pyjion.

File: Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst]: This text lands into https://docs.python.org/dev/whatsnew/changelog.html#changelog which is a changlog, you should not mention "a PR" there.

I suggest to just copy/paste what you wrote in whatsnew.

File: Doc/whatsnew/3.12.rst
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

**Review by vstinner — COMMENTED** [review]
Oh. This change reminds me my old https://peps.python.org/pep-0510/ nice :-)

File: Doc/whatsnew/3.12.rst
Comment (markshannon) [reply Doc/whatsnew/3.12.rst]: I'd rather not have a "purpose". We should state clearly what the function does, not bless some particular use case.

@adphrost
Could you add a warning that the new function pointer must have **exactly** the same behavior as the unaltered function.
I am a little concerned that C extensions will abuse this, and we will have deal with the bug reports.

**Comment by itamaro** [comment]
@vstinner @markshannon I think all comments were addressed - any further feedback?

File: Doc/c-api/function.rst
Comment (vstinner) [inline Doc/c-api/function.rst]:
```suggestion
   Set the vectorcall field of a given function object *func*.
```

File: Doc/whatsnew/3.12.rst
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

File: Doc/whatsnew/3.12.rst
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]: I don't think that this document is the right place to put a warning. Why this warning is not in the documentation of the function?

File: Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst]: Please remove this warning.

File: Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst]:
```suggestion
which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

File: Python/ceval.c
Comment (vstinner) [inline Python/ceval.c]: I have a personal preference for this coding style:

```suggestion
            if (Py_TYPE(function) == &PyFunction_Type
                && tstate->interp->eval_frame == NULL
                && ((PyFunctionObject *)function)->vectorcall == _PyFunction_Vectorcall) {
```

File: Objects/funcobject.c
Comment (vstinner) [inline Objects/funcobject.c:138]: @markshannon @brandtbucher @gvanrossum: I don't know if this change is correct, I don't know how _PyFunction_GetVersionForCurrentState() is supposed to work.

Comment (markshannon) [reply Objects/funcobject.c:138]: This is correct. `_PyFunction_GetVersionForCurrentState` returns 0 if it cannot set the version number to a valid version.

**Comment by itamaro** [comment]
@vstinner @markshannon thanks for the feedback. with @markshannon 's last response I think the remaining feedback is minor styling and phrasing. if that's all that left for this to be merged, I'm happy to perform these changes, but also totally ok with a core dev making these last changes when merging 😄

**Comment by markshannon** [comment]
One other thing.
To what extent does the VM have to respect the new vectorcall?

For example in `BINARY_SUBSCR_GETITEM` we do not check the function's version number, merely that it takes two arguments.
This saves us 4 bytes per `BINARY_SUBSCR` instruction.
So, if the `__getitem__` method were to have its vectorcall replaced, we might ignore it.

We should specify if, and when, the VM is obliged to respect the new vectorcall.
Is it OK if it is only for normal function calls, and not dunder methods?

**Comment by carljm** [comment]
@markshannon It would certainly be ideal if it could be respected for dunder methods too. If a dunder method is expensive and hot, we would want to be able to compile it.

I looked at the `BINARY_SUBSCR_GETITEM` code in main, and it looks like it currently does check the function version? Are you saying that you'd like to keep the option open to stop checking that in the future?

**Comment by carljm** [comment]
Is there a potential future where instead of doing version checks in the fast path (or missing changes entirely) we can use the function-modified hook that we've proposed to proactively invalidate inline caches on changes like this? (That's how Cinder inline caches are currently invalidated.)

**Comment by markshannon** [comment]
It looks like it is `LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN` that checks the argcount, so you will need to fix that to use the version number instead.

**Comment by carljm** [comment]
Discussed this in person. Sounds like we should probably audit ceval.c for specialized instructions (like `LOAD_ATTR_GETATTRIBUTE_OVERRIDEN`) that "inline" (jump to a function within ceval loop instead of going through the full call path) to make sure they deopt on function version changes. But if we ensure those are fixed in this PR, sounds like it is OK to always check function version going forward.

File: Python/ceval.c
Comment (Fidget-Spinner) [reply Python/ceval.c]: The author's code follows PEP 7 https://peps.python.org/pep-0007/#code-lay-out.

**Comment by itamaro** [comment]
pushed a fix to `LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN` to deopt on function version instead of argcount

@markshannon I also reviewed the interpreter loop and didn't found any other cases that inline calls without deopting on function version. the methodology I used is searching for occurrences of `inlined_py_calls` - let me know if this should cover everything!

(btw, I noticed that `BINARY_SUBSCR_GETITEM` contains `CALL_STAT_INC(inlined_py_calls)` twice - is that intentional?)

**Review by Fidget-Spinner — COMMENTED** [review]
The parts touching the specializer and specialized opcodes LGTM.

**Comment by Fidget-Spinner** [comment]
@itamaro 
> (btw, I noticed that BINARY_SUBSCR_GETITEM contains CALL_STAT_INC(inlined_py_calls) twice - is that intentional?)

No that's wrong.

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: PEP 7 has the brace on the following line

**Review by markshannon — COMMENTED** [review]
One minor formatting nit. Otherwise looks good.

File: Python/ceval.c
Comment (itamaro) [reply Python/ceval.c]: fixed

**Comment by itamaro** [comment]
> > (btw, I noticed that BINARY_SUBSCR_GETITEM contains CALL_STAT_INC(inlined_py_calls) twice - is that intentional?)
> 
> No that's wrong.

I'll file a separate issue for that

**Comment by markshannon** [comment]
All looks good to me.
@vstinner any thing else before I merge this?

File: Modules/_testcapi/vectorcall.c
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall.c]: This should be a TypeError.

```suggestion
        PyErr_SetString(PyExc_TypeError, "'func' must be a function");
```

File: Modules/_testcapi/vectorcall.c
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall.c]: PEP 7

```suggestion
override_vectorcall(PyObject *callable, PyObject *const *args, size_t nargsf,
                    PyObject *kwnames)
{
```

File: Objects/funcobject.c
Comment (erlend-aasland) [inline Objects/funcobject.c:219]: Might consider adding an `assert(func != NULL)` here for debugging convenience.

```suggestion
    assert(func != NULL);
    func->func_version = 0;
```

File: Python/ceval.c
Comment (erlend-aasland) [inline Python/ceval.c:3131]: Is it possible to add a test for this deoptimisation?

Comment (itamaro) [reply Python/ceval.c:3131]: is there an existing test I can use as a reference for how to trigger this specialization? I couldn't find any from a cursory search

Comment (Fidget-Spinner) [reply Python/ceval.c:3131]: This specialisation triggers with the following code:

```python

class X:
    def __getattribute__(self, attr):
        return attr

x = X()
for _ in range(8):
    x.a  #LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN
```

Comment (itamaro) [reply Python/ceval.c:3131]: thanks, I tried devising a test for this, PTAL

File: Lib/test/test_call.py
Comment (Fidget-Spinner) [inline Lib/test/test_call.py]: Could you set a `QUICKENING_WARMUP_DELAY` constant and use it here please? We should try to unify these values in another PR. See for example https://github.com/python/cpython/blob/main/Lib/test/test_dis.py#L761.

File: Lib/test/test_call.py
Comment (Fidget-Spinner) [inline Lib/test/test_call.py]: Please pull both `self.assertEqual` out of the loops. This itself is a `LOAD_ATTR` and there have been situations in the past where it confounded some of the test results when the test fails.

so
```python
assertEqual = self.assertEqual
...
```

File: Python/ceval.c
Comment (erlend-aasland) [reply Python/ceval.c:3131]: Thanks! I gave it a try, but unfortunately, the test does not fail if I remove the deopt.

File: Lib/test/test_call.py
Comment (Fidget-Spinner) [inline Lib/test/test_call.py]: Ditto about using `QUICKENING_WARMUP_DELAY` Maybe use `QUICKENING_WARMUP_DELAY * 5`

**Review by Fidget-Spinner — COMMENTED** [review]
The test LGTM in general but I have a few comments. Thanks!

File: Lib/test/test_call.py
Comment (itamaro) [reply Lib/test/test_call.py]: done

File: Lib/test/test_call.py
Comment (itamaro) [reply Lib/test/test_call.py]: used `*10` to be sure :)

File: Lib/test/test_call.py
Comment (itamaro) [reply Lib/test/test_call.py]: done

File: Python/ceval.c
Comment (itamaro) [reply Python/ceval.c:3131]: thanks for checking @erlend-aasland ! 

I realized I was testing something else (that specialization doesn't trigger for an already-overridden vectorcall)

pushed an update that now tests both scenarios (avoid specialization, and deopt) - verified the `test_setvectorcall_load_attr_specialization_deopt` fails when removing the deopt

File: Lib/test/test_call.py
Comment (erlend-aasland) [reply Lib/test/test_call.py]: We should consider moving `QUICKENING_WARMUP_DELAY` to `test.support` in a follow-up PR.

File: Python/ceval.c
Comment (erlend-aasland) [reply Python/ceval.c:3131]: Thanks, looks good! :)

**Review by erlend-aasland — APPROVED** [review]

**Comment by gvanrossum** [comment]
Almost ready. Who's on the hook to get this merged?

**Review by markshannon — APPROVED** [review]
Looks good now.

**Comment by markshannon** [comment]
I'll merge once the conflict has been resolved.

**Comment by vstinner** [comment]
PyFunction_SetVectorcall() is really a cool feature, I love it! I'm not sure yet how people will manage to abuse it, but I love it :-D

**Pull request merged**
